File: app/src/main/java/com/waveneuro/ui/user/password/password/confirm/SetNewPasswordActivity.java
package com.waveneuro.ui.user.password.password.confirm;

import android.graphics.Color;
import android.os.Bundle;
import android.text.Editable;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.TextWatcher;
import android.text.style.ForegroundColorSpan;
import android.text.style.UnderlineSpan;
import android.view.View;
import android.view.Window;
import android.widget.TextView;

import androidx.lifecycle.Observer;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.snackbar.Snackbar;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import com.waveneuro.R;
import com.waveneuro.ui.base.BaseFormActivity;
import com.waveneuro.ui.dashboard.DashboardCommand;
import com.waveneuro.ui.user.login.LoginCommand;

import java.util.regex.Pattern;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class SetNewPasswordActivity extends BaseFormActivity {

    @BindView(R.id.et_password)
    TextInputLayout etPassword;

    @BindView(R.id.tip_password)
    TextInputEditText tipPassword;

    @BindView(R.id.btn_reset_password)
    MaterialButton btnResetPassword;

    @BindView(R.id.tv_condition1)
    TextView tvCondition1;

    @BindView(R.id.tv_condition2)
    TextView tvCondition2;

    @BindView(R.id.tv_condition3)
    TextView tvCondition3;

    @BindView(R.id.tv_condition4)
    TextView tvCondition4;

    @BindView(R.id.tv_log_in)
    TextView tvLogin;

    @Inject
    LoginCommand loginCommand;

    @Inject
    DashboardCommand dashboardCommand;

    @Inject
    SetNewPasswordViewModel setNewPasswordViewModel;

    private String email;
    private String code;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        getWindow().requestFeature(Window.FEATURE_ACTIVITY_TRANSITIONS);
        getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN);
        getWindow().setStatusBarColor(Color.TRANSPARENT);

        super.onCreate(savedInstanceState);
        activityComponent().inject(this);
        setContentView(R.layout.activity_set_new_password);
        ButterKnife.bind(this);

        if (getIntent().hasExtra(SetNewPasswordCommand.EMAIL)
                && getIntent().hasExtra(SetNewPasswordCommand.CODE)) {
            email = getIntent().getStringExtra(SetNewPasswordCommand.EMAIL);
            code = getIntent().getStringExtra(SetNewPasswordCommand.CODE);
        }
        setView();
        setObserver();

        tipPassword.addTextChangedListener(mTextEditorWatcher);

        this.setNewPasswordViewModel.processEvent(new SetNewPasswordViewEvent.Start());
    }

    private final TextWatcher mTextEditorWatcher = new TextWatcher() {
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        public void onTextChanged(CharSequence s, int start, int before, int count) {
            boolean isLong = s.length()>=8;
            boolean lowerAndUpper = Pattern.compile("[A-Z]").matcher(s).find() && Pattern.compile("[a-z]").matcher(s).find();
            boolean hasNumbers =  Pattern.compile("[0-9]").matcher(s).find();
            boolean hasSpecials =  Pattern.compile("[^a-zA-Z0-9]").matcher(s).find();

            tvCondition1.setTextColor(getResources().getColor(isLong?R.color.aqua:R.color.red));
            tvCondition2.setTextColor(getResources().getColor(lowerAndUpper?R.color.aqua:R.color.red));
            tvCondition3.setTextColor(getResources().getColor(hasNumbers?R.color.aqua:R.color.red));
            tvCondition4.setTextColor(getResources().getColor(hasSpecials?R.color.aqua:R.color.red));

            boolean isCorrect = isLong && lowerAndUpper && hasNumbers && hasSpecials;

            btnResetPassword.setClickable(isCorrect);
            btnResetPassword.setBackgroundColor(getResources().getColor(isCorrect?R.color.aqua:R.color.gray));

        }

        public void afterTextChanged(Editable s) {
            int x = 0;
        }
    };

    private void setView() {
        logInSpanText();
    }

    private void logInSpanText() {
        SpannableString spannableString = new SpannableString(getString(R.string.log_in));
        spannableString.setSpan(new UnderlineSpan(), 0, 6, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
        spannableString.setSpan(new ForegroundColorSpan(getResources().getColor(R.color.yellow_dim)), 0, 6, Spannable.SPAN_EXCLUSIVE_INCLUSIVE);
        tvLogin.setText(spannableString);
    }

    private void setObserver() {
        this.setNewPasswordViewModel.getData().observe(this, setNewPasswordViewStateObserver);
        this.setNewPasswordViewModel.getViewEffect().observe(this, setNewPasswordViewEffectObserver);
    }

    Observer<SetNewPasswordViewState> setNewPasswordViewStateObserver = viewState -> {
        if (viewState instanceof SetNewPasswordViewState.Success) {
            SetNewPasswordViewState.Success success = (SetNewPasswordViewState.Success) viewState;

            View view = findViewById(R.id.root);
            final Snackbar snackBar = Snackbar.make(view, R.string.password_has_been_changed, Snackbar.LENGTH_LONG);
            snackBar.setDuration(100000000);
            snackBar.setAction(R.string.ok, new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    snackBar.dismiss();
                    loginCommand.navigate();
                }
            });
            snackBar.show();
        } else if (viewState instanceof SetNewPasswordViewState.Failure) {
            SetNewPasswordViewState.Failure error = (SetNewPasswordViewState.Failure) viewState;
            onFailure(error.getError());
        } else if (viewState instanceof SetNewPasswordViewState.Loading) {
            SetNewPasswordViewState.Loading loading = ((SetNewPasswordViewState.Loading) viewState);
            if (loading.getLoading())
                displayWait("Loading...", null);
            else
                removeWait();
        }
    };

    Observer<SetNewPasswordViewEffect> setNewPasswordViewEffectObserver = viewEffect -> {
        if (viewEffect instanceof SetNewPasswordViewEffect.Home) { }
    };

    @OnClick(R.id.btn_reset_password)
    public void onClickSubmitNewPassword() {
        submit();
    }

    @OnClick(R.id.tv_log_in)
    public void onClickLogIn() {
        loginCommand.navigate();
    }

    @Override
    public void submit() {
        this.setNewPasswordViewModel.processEvent(
                new SetNewPasswordViewEvent.SetNewPassword(
                        email,
                        code,
                        tipPassword.getText().toString()
                ));
    }
}

File: app/src/main/java/com/waveneuro/ui/user/password/new_password/SetNewPasswordActivity.kt
package com.waveneuro.ui.user.password.new_password

import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.text.*
import android.text.style.ForegroundColorSpan
import android.text.style.UnderlineSpan
import android.view.View
import android.view.Window
import androidx.activity.viewModels
import androidx.lifecycle.Observer
import com.google.android.material.snackbar.Snackbar
import com.waveneuro.R
import com.waveneuro.databinding.ActivitySetNewPasswordBinding
import com.waveneuro.ui.base.activity.BaseViewModelActivity
import com.waveneuro.ui.dashboard.DashboardCommand
import com.waveneuro.ui.user.login.LoginActivity
import com.waveneuro.ui.user.password.new_password.viewmodel.SetNewPasswordViewModel
import com.waveneuro.ui.user.password.new_password.viewmodel.SetNewPasswordViewModelImpl
import com.waveneuro.utils.ext.getAppComponent
import java.util.regex.Pattern
import javax.inject.Inject

class SetNewPasswordActivity :
    BaseViewModelActivity<ActivitySetNewPasswordBinding, SetNewPasswordViewModel>() {

    @Inject
    var dashboardCommand: DashboardCommand? = null

    override val viewModel: SetNewPasswordViewModelImpl by viewModels {
        getAppComponent().setNewPasswordViewModelFactory()
    }

    private var email: String? = null
    private var code: String? = null

    override fun initBinding(): ActivitySetNewPasswordBinding =
        ActivitySetNewPasswordBinding.inflate(layoutInflater)

    override fun onCreate(savedInstanceState: Bundle?) {
        window.requestFeature(Window.FEATURE_ACTIVITY_TRANSITIONS)
        window.decorView.systemUiVisibility = (View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN)
        window.statusBarColor = Color.TRANSPARENT
        super.onCreate(savedInstanceState)

        if (intent.hasExtra(EMAIL)
            && intent.hasExtra(CODE)
        ) {
            email = intent.getStringExtra(EMAIL)
            code = intent.getStringExtra(CODE)
        }
        setView()
        setObserver()
        binding.tipPassword.addTextChangedListener(mTextEditorWatcher)
    }

    private val mTextEditorWatcher: TextWatcher = object : TextWatcher {
        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            val isLong = s.length >= 8
            val lowerAndUpper =
                Pattern.compile("[A-Z]").matcher(s).find() && Pattern.compile("[a-z]").matcher(s)
                    .find()
            val hasNumbers = Pattern.compile("[0-9]").matcher(s).find()
            val hasSpecials = Pattern.compile("[^a-zA-Z0-9]").matcher(s).find()
            val isCorrect = isLong && lowerAndUpper && hasNumbers && hasSpecials
            with(binding) {
                tvCondition1.setTextColor(resources.getColor(if (isLong) R.color.aqua else R.color.red))
                tvCondition2.setTextColor(resources.getColor(if (lowerAndUpper) R.color.aqua else R.color.red))
                tvCondition3.setTextColor(resources.getColor(if (hasNumbers) R.color.aqua else R.color.red))
                tvCondition4.setTextColor(resources.getColor(if (hasSpecials) R.color.aqua else R.color.red))
                btnResetPassword.isClickable = isCorrect
                btnResetPassword.setBackgroundColor(resources.getColor(if (isCorrect) R.color.aqua else R.color.gray))
            }
        }

        override fun afterTextChanged(s: Editable) {}
    }

    private fun setView() {
        with(binding) {
            btnResetPassword.setOnClickListener {
                submit()
            }
            tvLogIn.setOnClickListener {
                startActivity(LoginActivity.newIntent(this@SetNewPasswordActivity))
            }
        }
        logInSpanText()
    }

    private fun setObserver() {
        viewModel.viewEffect.observe(this, Observer { viewEffect ->
            when (viewEffect) {
                is SetNewPasswordViewEffect.Success -> {
                    showSnackBar()
                }
            }
        })
    }

    private fun logInSpanText() {
        val spannableString = SpannableString(getString(R.string.log_in))
        spannableString.setSpan(UnderlineSpan(), 0, 6, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
        spannableString.setSpan(
            ForegroundColorSpan(resources.getColor(R.color.yellow_dim)),
            0,
            6,
            Spannable.SPAN_EXCLUSIVE_INCLUSIVE
        )
        binding.tvLogIn.text = spannableString
    }

    private fun showSnackBar() {
        val snackBar =
            Snackbar.make(binding.root, R.string.password_has_been_changed, Snackbar.LENGTH_LONG)
        snackBar.duration = 100000000
        snackBar.setAction(R.string.ok) {
            snackBar.dismiss()
            startActivity(LoginActivity.newIntent(this@SetNewPasswordActivity))
        }
        snackBar.show()
    }

    private fun submit() {
        viewModel.processEvent(
            SetNewPasswordViewEvent.SetNewPassword(
                email ?: "",
                code ?: "",
                binding.tipPassword.text.toString()
            )
        )
    }

    companion object {
        private const val EMAIL = "email"
        private const val CODE = "code"

        fun newIntent(context: Context, email: String?, code: String?) = Intent(
            context,
            SetNewPasswordActivity::class.java
        ).apply {
            flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
            putExtra(EMAIL, email)
            putExtra(CODE, code)
        }
    }
}